src/backups.py

```diff
@@ -0,0 +1,307 @@
+            s3_parameters["secret-key"],
+        )
+        if not success:
+            logger.warning(error_message)
+            event.fail(error_message)
+
+            success, error_message = self._post_backup()
+            if not success:
+                logger.warning(error_message)
+
+            return
+
+        # Run operations to clean up after the backup
+        success, error_message = self._post_backup()
+        if not success:
+            logger.warning(error_message)
+            event.fail(error_message)
+            return
+
+        event.set_results(
+            {
+                "backup-id": datetime_backup_requested,
+            }
+        )
+
+    def _retrieve_s3_parameters(self) -> Tuple[Dict, List[str]]:
+        """Retrieve S3 parameters from the S3 integrator relation."""
+        s3_parameters = self.s3_integrator.get_s3_connection_info()
+
+        required_parameters = [
+            "bucket",
+            "access-key",
+            "secret-key",
+        ]
+        missing_required_parameters = [
+            param for param in required_parameters if param not in s3_parameters
+        ]
+        if missing_required_parameters:
+            logger.warning(
+                f"Missing required S3 parameters in relation with S3 integrator: {missing_required_parameters}"
+            )
+            return {}, missing_required_parameters
+
+        # Add some sensible defaults (as expected by the code) for missing optional parameters
+        s3_parameters.setdefault("endpoint", "https://s3.amazonaws.com")
+        s3_parameters.setdefault("region")
+        s3_parameters.setdefault("path", "")
+
+        return s3_parameters, []
+
+    def _can_unit_perform_backup(self) -> Tuple[bool, str]:
+        """Validates whether this unit can perform a backup."""
+        logger.info("Checking if cluster is in blocked state")
+        if self.charm._is_cluster_blocked():
+            return False, "Cluster or unit is in a blocking state"
+
+        logger.info("Checking state and role of unit")
+
+        try:
+            state, role = self.charm._mysql.get_member_state()
+        except MySQLGetMemberStateError:
+            return False, "Error obtaining member state"
+
+        if role == "primary" and self.charm.app.planned_units() > 1:
+            return False, "Unit cannot perform backups as it is the cluster primary"
+
+        if state in ["recovering", "offline", "error"]:
+            return False, f"Unit cannot perform backups as its state is {state}"
+
+        return True, None
+
+    def _pre_backup(self) -> Tuple[bool, str]:
+        """Runs operations required before performing a backup."""
+        logger.info("Setting cluster state as 'backing-up'")
+        self.charm.unit_peer_data["cluster-state"] = "backing-up"
+
+        try:
+            logger.info("Setting unit option tag:_hidden")
+            self.charm._mysql.set_instance_option("tag:_hidden", "true")
```

Comment:
we should do this only `if charm.app.planned_units() > 1`

Reply:
addressed in [dcce74c](https://github.com/canonical/mysql-k8s-operator/pull/136/commits/dcce74c6c995024fd3cae92853903eb11cf09399)